
# SIFIRDAN PYTHON ÖĞRENME SÜRECİ


Jupyter Notebook üzerindeki bilinmesi gereken bazı kısa yol tuşları;

Shift + Enter = Seçili olan hücredeki (Cell) kodu çalıştırır ve altındaki hücreye geçer.

Alt + Enter = Seçili olan hücreyi çalıştırır ve altına yeni bir boş hücre açar.

CTRL + Enter = Sadece seçili olan hücreyi çalıştırır ve durur.

"#" = Markdown tipindeki hücreye diez işaretini koyduğunuzda başlık yapmanıza olanak tanır. Code tipindeki hücre içerisinde diez kullandığınızda ise o satırı metinsel ifade (açıklama) olarak görür ve koda dahil etmez.

Hücre tipini (cell type) belirleme için aşağıdaki kısayol tuşları kullanılır;

Y = "Code" Kod yazmak için kullanılır.

M = "Markdown" Başlık, açıklama vb. metinsel ifadeler için kullanılır.

R = "Raw NBConvert" 

NOT: Aşağıda yer alan kısayol tuşlarının çalışması için hücre içerisinin seçili olmaması gerekir. Yani kod yazılan satıra değil "In, Out" bölümüne tıklanarak kısayol tuşları çalıştırılmalıdır. İlgili yere tıkladığınızda hücrenin mavi renge döndüğünü fark edeceksiniz. Yeşil renk iken basmış olduğunuz tuşlar hücre içerisine yazılacaktır.

D+D = Herhangi bir hücre seçiliyken iki kere D tuşuna basılırsa o hücreyi siler.

A = Seçili hücrenin üstüne yeni hücre ekler.

X = Seçili olan hücreyi keser.

C = Seçili hücreyi kopyalar.

V = Kopyalanan hücreyi seçilen hücrenin altına yapıştırır.

Shift + V = Kopyalanan hücreyi seçilen hücrenin üzerine yapıştırır.

Z = Silinen hücreyi geri getirir

CTRL + S = Kayıt işlemi yapar.


# DERS-1


# Sayılar ve Aritmetik

Bu üç sayı tipiyle öntanımlı olarak yedi aritmetik işlem yapılabilir: Toplama (+), çıkarma (-), çarpma (*), bölme (/), tamsayı bölme (//), kalan (%), ve üs alma (**).

In [1]:
2+3

5

In [2]:
7-5, 5-7

(2, -2)

In [3]:
23+5

28

In [4]:
3**254 # üst alma işlemlerinde kullanılır. Tamsayılarda üst sınır yoktur; bellek yettiğince büyük sayılara çıkılabilir.

15445383597460525862771087421073478123802094534281015332400536047781785847751826215666863239345512169090185060060988826169

In [5]:
2.9**1000 #Reel sayılarda ise üst sınırı aşma (overflow) problemi, başka dillerde olduğu gibi mevcuttur.

OverflowError: (34, 'Result too large')

In [6]:
10/3

3.3333333333333335

In [7]:
10//3 # Tam sayı bölme işlemi yaptığı için virgülden sonraki kısmın çıktısını vermez.

3

In [8]:
10%3 # (Remainder) 10/3 işleminden kalanı gösterir. Genellikle sayıların çift mi tek mi olduğunu belirlemede kullanılır.

1


# Değişken Atamaları

Python'da değişkenlerin tiplerini (int, float, vs.) önceden beyan etmek gerekmez. Değişken isimleri, atamanın sağ tarafındaki ifadeye işaret eden bir isimdir sadece.

In [9]:
x = -2      # bir tamsayı
y = 3/4 * 2 # reel sayı
z = "Merhaba"+"Dünya" # string
x,y,z # Yukarıda oluşturulan 3 değişkeni de yazdırdık.

(-2, 1.5, 'MerhabaDünya')

In [10]:
x=5 
y=10 
x*y

50

Çoklu atamalar yapmak mümkündür. 

In [11]:
x, y, z = -2, 1.5, "Merhaba Dünya"
x,y,z

(-2, 1.5, 'Merhaba Dünya')

Bu usulü kullanarak iki değişkenin değerini, bir ara değişken kullanmadan değiştokuş etmek mümkün olur.

In [12]:
x,y = y,x # Kodu çalıştırmadan önce x=-2, y=1.5 idi. Bu kod ile x'i y değerine y'yi ise x değerine eşitledik ve böylece artık
x,y       # x=1.5, y=-2 olmuş oldu.

(1.5, -2)

Python 3 ile değişken isimlerinde herhangi Unicode karakterleri kullanabilirsiniz. Yani "ı,o,ğ,ç" gibi harflerin kullanımına izin verir. Yine de bu durum kodların başka Python versiyonlarında da çalışması adına pek tavsiye edilmez.

In [13]:
kağıt_sayısı, ölçü = 5, 0.5
kağıt_sayısı, ölçü

(5, 0.5)

Bir değişkeni del komutuyla bellekten silebiliriz. Genel olarak, belli bir isme bağlı bir nesneyi bellekten kaldırır. Sıralı nesnelerde, bir dilimi silmek için de kullanılabilir.

In [14]:
del ölçü
ölçü

NameError: name 'ölçü' is not defined

In [15]:
L = [1,2,3,4,5,6]
del L[1]   # ikinci elemanı siler
L

[1, 3, 4, 5, 6]

In [16]:
del L[2:4]    # üçüncüden dördüncüye kadar olan elemanları siler
L

[1, 3, 6]

In [17]:
del L[:]      # bütün elemanları siler, ama L listesini silmez.
L

[]

In [18]:
del L         # L listesini siler
L      # Listeyi sildik ve ekrana yazmak istediğimizde böyle bir ifade bulamadığı için hata verdi

NameError: name 'L' is not defined

# Integer, Float ve Complex

Tam sayılar integer (int), kesirli sayılar float, karmaşık sayılar ise complex olarak tanımlanır. Herhangi bir değişkenin tipini öğrenmek için "type" komutu kullanılır.


In [19]:
ilkDeger = 10
ikinciDeger = 20
toplam = ilkDeger + ikinciDeger
bolum = ilkDeger/ikinciDeger

In [20]:
toplam, type(toplam)

(30, int)

In [21]:
bolum, type(bolum)

(0.5, float)

İki farklı tipte sayı aynı işlemde kullanılırsa, hangi tip daha genelse sonuç o tipte verilir. Tamsayı ve reel sayı kullanılırsa sonuç reeldir; reel sayı ve karmaşık sayı kullanılırsa sonuç karmaşıktır.
Herhangi bir değişkenin tipini öğrenmek için "type" komutu kullanılır.

In [22]:
type(1+4), type(1+4.0), type(1+4+0j)

(int, float, complex)

# Öntanımlı Veri Yapıları


# String

Metinsel ifadeler string olarak tanımlanır. Herhangi bir değişkeni string (str) olarak tanımladığımızda içerisinde rakam olsa bile bilgisayar onu metin olarak algılayacaktır.

In [23]:
ilkMetinselRakam = str(10)
ikinciMetinselRakam = str(20)
metinselToplam = ilkMetinselRakam+ikinciMetinselRakam
metinselToplam

'1020'

Yukarıdaki örnekte gördüğümüz gibi string olarak tanımladığımız için 10+20 sonucunu 30 olarak değilde 1020 olarak verdi. 
Çünkü o bilgisayara göre artık bir metin. Type komutu ile tipini kontrol ettiğimizde string olduğunu görebiliriz;

In [24]:
type(metinselToplam)

str

Aynı şekilde kullanıcıdan rakamsal olarak girdi almak istediğimizde bunu integer olarak tanımlamamız gerekir. Aksi halde aşağıdaki örnekte olduğu gibi metinsel ifade olarak algılayacaktır. 

In [28]:
ilkGirdi = input("Bir rakam giriniz = ")

Bir rakam giriniz = 10


In [29]:
ilkGirdi+10

TypeError: can only concatenate str (not "int") to str

Şimdi ise kullanıcıdan alınan veriyi int olarak tanımlayalım ve sonucu görelim;

In [30]:
ikinciGirdi = int(input("Bir rakam giriniz = "))

Bir rakam giriniz = 10


In [31]:
ikinciGirdi+10

20

### Index ve Slicing

Index yardımı ile string, liste, tuple, dictionary gibi verilerin istediğimiz elamanını veya karakterini çağırabiliriz
Aynı şekilde veri içerisindeki istediğimiz bölümü dilimleyebiliriz (slicing);

In [32]:
isim = "Enes YILMAZ"  # Bu bir metinsel ifadedir (string). Şimdi bu ifadenin istediğimiz elamınını çağırabiliriz.
isim[0] # İlk karakterin indexi sıfırdan başlamaktadır. 

'E'

In [33]:
isim[1] # İkinci karakter

'n'

In [34]:
isim[-1] # Son karakter

'Z'

In [35]:
isim[-2] # Sondan bir önceki karakter

'A'

In [36]:
isim[::] # Tamamını çağırmak için kullanılır.

'Enes YILMAZ'

In [37]:
isim[1:3] # Kaçıncı karakterden başlayacağını ve nerde duracağını seçebiliriz. 3'ü dahil etmez.

'ne'

In [38]:
isim[0:10:3] # Başlangıç, bitiş ve atlanacak karakter sayısını belirleyebiliriz.

'EsIA'

In [39]:
isim[3:] # İlk 3 karakteri atlar ve devamını alır.

's YILMAZ'

In [40]:
isim[:3] # Sadece ilk 3 karakteri alır.

'Ene'

In [41]:
isim[::-1] # Karakterlerin tamamını alır fakat sonra -1 olduğu için tersten başlar. Böylece metni kolayca ters çevirebiliriz.

'ZAMLIY senE'

In [42]:
isim[::-2] # Aynı şekilde karakterleri alırken atlama yapabiiriz.

'ZMI eE'

In [43]:
isim[0]+isim[5] # Birden fazla index kullanarak birleştirme yapılabilir. Direk topladığımız için yan yana yazdık. 

'EY'

In [44]:
isim[0]+ "."+isim[5] # İstersek bu şekilde alternatifler üretebiliriz.

'E.Y'

### Büyük-küçük harf değişiklikleri

Dizeleri tamamen küçük veya büyük harfe çevirmek, cümlenin ilk harfinin büyük olmasını sağlamak, veya bir başlıktaki gibi her kelimenin büyük harfle başlamasını sağlamak için çeşitli dize metodları vardır.

In [45]:
s = "Pijamalı hasta yağız şoföre çabucak güvendi."

print("s.upper():", s.upper())   # Hepsini büyük harfe çevir.
print("s.lower():", s.lower())   # Hepsini küçük harfe çevir.
print("s.capitalize():", s.capitalize())   # Sadece cümlelerin ilk harfini büyük yap. 
print("s.swapcase():", s.swapcase())   # Büyük harfleri küçüğe, küçükleri büyüğe çevir.
print("s.title():", s.title())   # Her kelimenin ilk harfini büyük yap.

s.upper(): PIJAMALI HASTA YAĞIZ ŞOFÖRE ÇABUCAK GÜVENDI.
s.lower(): pijamalı hasta yağız şoföre çabucak güvendi.
s.capitalize(): Pijamalı hasta yağız şoföre çabucak güvendi.
s.swapcase(): pIJAMALI HASTA YAĞIZ ŞOFÖRE ÇABUCAK GÜVENDI.
s.title(): Pijamalı Hasta Yağız Şoföre Çabucak Güvendi.


### Center, Split

In [46]:
isim.capitalize() # capitalize komutu ile metnin ilk harfini büyük kalanları küçük hale çevirebiliriz.

'Enes yilmaz'

In [47]:
isim # Fakat bunu başka bir değişene tanımalamadığımız sürece orjinal halini korumaya devam eder.

'Enes YILMAZ'

In [48]:
isim.center(25, '*') # Metni belirlenen karakter sayısına göre ortalar ve boşlukları ne ile dolduracağımızı seçebiliriz.

'*******Enes YILMAZ*******'

In [49]:
isim.split() # Metin içerisindeki her kelimeyi ayırır (boşluktan sonra)

['Enes', 'YILMAZ']

# Dize

Bir veya daha fazla karakterden bir araya gelmiş bir sıralı veri tipidir. Dizelerle düz metin halindeki veriler alınıp işlenebilir. Veriler pek çok zaman düz metin dosyaları halinde bulunur. HTML, XML, JSON veya CSV gibi standartlardaki veriler de düz metinden oluşur. Bir metin verisini dize halinde aldığımızda, dize metodlarını kullanarak bu verileri işleyebilir, biçimlerini değiştirebilir, parçalayabilir, birleştirebilir, sıralayabilir, değişiklik yapabiliriz.

## Temel işlemler

Sıralı nesnelerle kullanılan len, in, not in, +, * gibi işlemler dizelerde de aynı şekilde çalışır:

In [50]:
print( "Def" in "Abcç Defg" )  # Eğer "Def" ifadesi "Abcç Defg" nin içerisinde varsa True yoksa False döndürür.
print( "def" in "ABcç Defg" )  # Burada False dönmesinin sebebi büyük küçük karakter farkıdır.
print( "def" not in "ABcç Defg" ) # "not in" komutunu kullandığımızda metnin içerisinde yoksa True varsa False döndürür.

True
False
True


In [51]:
len("Enes YILMAZ") # "len" komutu ile uzunluk (karakter sayısı) hesaplanabilir. Boşlukları da dahil eder.

11

In [52]:
"Enes" + "Yılmaz"  # İstediğimiz kadar string tipindeki veriyi toplayabiliriz.

'EnesYılmaz'

In [53]:
"Enes" + " " + "YILMAZ"

'Enes YILMAZ'

In [54]:
"Enes YILMAZ" + 5  # Bir integer ile toplama yaptığımızda hata alırız.

TypeError: can only concatenate str (not "int") to str

In [55]:
"Enes YILMAZ" * 5 # Fakat integer bir değişken ile çarpma işlemi yapabiliriz. 

'Enes YILMAZEnes YILMAZEnes YILMAZEnes YILMAZEnes YILMAZ'

In [56]:
("Enes YILMAZ" + " ")*5

'Enes YILMAZ Enes YILMAZ Enes YILMAZ Enes YILMAZ Enes YILMAZ '

## Tek tırnak, çift tırnak, üçlü tırnak

Python'da diziler istenirse tek tırnakla, istenirse çift tırnakla sınırlandırarak kullanılabilirler. İkisi de aynı nesneyi üretir; yani "abc" ile 'abc' aynı dizedir.

Bu esneklik, dizenin içinde bir çeşit tırnak varsa, dizenin çevresinde öbür çeşit tırnağı kullanma imkânı verir.

In [57]:
print("Enes'de bizimle gelecek mi?")

Enes'de bizimle gelecek mi?


In [58]:
print('Dedim ki: "Gel!"')

Dedim ki: "Gel!"


Eğer dize içerisinde her iki tırnak (",') işaretini de kullanmamaız gerekiyorsa, kaçış karakterleri (\) kullanabiliriz. Tırnak açtıkdan sonra "\" (ters slash) işaretini koyduğumuzda kendinden sonra gelen karakteri kaçış karakteri olarak algılar ve metinsel ifade olarak yazar.

In [59]:
print("Enes'e dedim ki: \"Gelme!\"")

Enes'e dedim ki: "Gelme!"


Üç tırnak (""" veya ''') kullanarak birkaç satıra yayılan bir dize yaratabiliriz. Yaptığımız kaydırmalar, satırbaşları, boş satırlar dizenin parçası olarak kaydedilir. Bu tür dizeler özellikle fonksiyonlardaki belge dizelerinde (docstring) çok kullanılır.

In [60]:
ucTirnak = """
              Enes
              YILMAZ
              1996"""
print(ucTirnak)


              Enes
              YILMAZ
              1996


## Dize metodları

Dize sınıfı içinde tanımlanmış pek çok metod mevcut. Bunların tam bir listesini görmek için dir(str) komutunu verebilirsiniz.

In [61]:
dir(str)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

Her bir metodun kısa bir açıklaması için help() komutunu kullanabilirsiniz.

In [62]:
help(str.translate)

Help on method_descriptor:

translate(self, table, /)
    Replace each character in the string using the given translation table.
    
      table
        Translation table, which must be a mapping of Unicode ordinals to
        Unicode ordinals, strings, or None.
    
    The table must implement lookup/indexing via __getitem__, for instance a
    dictionary or list.  If this operation raises LookupError, the character is
    left untouched.  Characters mapped to None are deleted.



Burada bu metodların belli başlılarının kullanımlarını örnekleyerek açıklayacağız. Özellikle, metin şeklindeki verilerin analizinde kullanılabilecek metodları seçtik. Tam bir liste için [Python yardım belgelerine](https://docs.python.org/3.6/library/stdtypes.html#string-methods)  bakabilirsiniz.

startswith, endswith
----------------

Bir dizenin başlangıcında veya sonunda belli bir alt dizenin bulunması halinde `True` verirler.

In [63]:
s = "Python Programlamaya Giriş"
s.startswith("Pyth"), s.startswith("python")

(True, False)

In [64]:
s.endswith("iriş"), s.endswith("xyz")

(True, False)

Bir alt dizenin yerini bulmak: find, rfind, index, rindex
-------------------
`find` metodu, bir dizenin içinde bir alt dizenin ilk bulunduğu konumu verir. Aranan alt dize mevcut değilse -1 döndürür.

In [65]:
s = "ABRAKADABRA"  # Aradığımız metin 'RA' olduğu için ilk gördüğü 'RA' nın başlangıç indexini verir.
s.find("RA")

2

In [66]:
s.find("XYZ")  # Bulamadığı zaman -1 döndürür.

-1

Konum saymanın sıfır ile başladığını hatırlayın; o yüzden "RA" dizesi üçüncü harfte başlıyor, ama konum indeksi 2.

Dikkat: Amacınız belli bir alt dizenin yerini bulmak değil, sadece mevcut olup olmadığını anlamaksa find yerine in kullanın.

In [67]:
"RA" in s, "XYZ" in s # Aradığımız ifade "s" değişkeninin içerisinde varsa True yoksa False döndürür.

(True, False)

Aramaya sağdan başlamak isterseniz rfind metodunu kullanın.

In [68]:
s.rfind("RA") # İfadenin sonundan başlarak ilk gördüğü 'RA' nın indexini verir.

9

index ile find aynı işi yapar. Tek farkları, aranan alt dize bulunamayınca index'in bir hata işareti ("ValueError") vermesidir. Hata yakalamaya dayalı programlarda tercih edilir.

In [69]:
s.index("RA")

2

In [70]:
s.index("XYZ")

ValueError: substring not found

Aramaya sağdan başlamak isterseniz `rindex` komutunu kullanabilirsiniz.

Dizeleri birleştirmek: join
-------------
Bir sıralı nesnenin (liste, çokuz, dize, vs.) dize tipi elemanlarını birleştirerek yeni bir dize üretir.

In [71]:
"".join(["A","67","B","0.15"])    # yanyana yapıştırır

'A67B0.15'

In [72]:
" ".join(["A","67","B","0.15"])   # araya bir boşluk koyarak yapıştırır

'A 67 B 0.15'

In [73]:
"-*-".join(["A","67","B","0.15"]) # araya -*- koyarak yapıştır

'A-*-67-*-B-*-0.15'

In [74]:
L = ['m', 'e', 'r', 'h', 'a', 'b', 'a']
"".join(L)

'merhaba'

Boşlukları atmak: lstrip, rstrip, strip
--------------------
Bu metodlar bir dizenin sol ve/veya sağ tarafındaki boş karakterleri (boşluk, kaydırma `\t`, satırbaşı `\n`) kaldırır. `lstrip` soldaki, `rstrip` sağdaki, `strip` ise her iki yandaki boşlukları temizler. Bu metodlar özellikle bir dosya satır satır okunurken yararlı olur.

In [75]:
s = """
       Enes YILMAZ   
"""
s

'\n       Enes YILMAZ   \n'

In [76]:
s.lstrip()  # Dizenin solunda kalan boşlukları siler.

'Enes YILMAZ   \n'

In [77]:
s.rstrip()  # Dizenin sağında halan boşlukları siler.

'\n       Enes YILMAZ'

In [78]:
s.strip()  # Dizenin hem sağ hem de sol tarafındaki boşluklarını siler.

'Enes YILMAZ'

Değiştirme: replace
-----------
Dizede belli bir alt dizenin ortaya çıktığı her yeri başka bir alt dizeyle değiştirir.

In [79]:
s = "eskişehir'de eski bir evde eski bir kapı"
s.replace("eski", "yeni")

"yenişehir'de yeni bir evde yeni bir kapı"

Tercihe bağlı olarak, soldan itibaren sadece belli sayıda değiştirme de yapılabilir:

In [80]:
s.replace("eski", "yeni", 2)

"yenişehir'de yeni bir evde eski bir kapı"

Bölme: split, rsplit, splitlines
----------------
`split` metodu bir dizedeki kelimeleri belli yerlerden (varsayılan olarak boşluk karakterlerinden) böler ve ortaya çıkan bölümlerden oluşan bir dize listesi verir. Fazlalık boşluklar hesaba katılmaz.

In [81]:
"   kedi köpek   balık    kuş   ".split()

['kedi', 'köpek', 'balık', 'kuş']

Boşluk yerine başka bir ayırıcı kullanılabilir. Ayrıcı istenilen sayıda karakterden oluşabilir.

In [82]:
"1,2,3,4,5,6".split(",")

['1', '2', '3', '4', '5', '6']

In [83]:
"a<>b<>c<>d<>e".split("<>")

['a', 'b', 'c', 'd', 'e']

Sadece belli sayıda bölme yapmak da mümkündür. Arta kalan kısım tek bir dizede toplanır.

In [84]:
"   kedi köpek   balık    kuş   ".split(maxsplit=2)

['kedi', 'köpek', 'balık    kuş   ']

Soldan değil sağdan başlayarak ayırmamız gerekirse rsplit metodunu kullanırız.

In [85]:
"   kedi köpek   balık    kuş   ".rsplit(maxsplit=2)

['   kedi köpek', 'balık', 'kuş']

Birden fazla satır barındıran bir dizeyi satır sınırlarından bölmek için splitlines metodu kullanılabilir.

In [86]:
s = """abcç
defgğ
hıi
jkl"""

s.splitlines()

['abcç', 'defgğ', 'hıi', 'jkl']

splitlines metodu sadece satırbaşı ("\n") ile değil, "\r", "\v", "\f", ve satır sonu sayılan başka birkaç karakter ile de böler. Ayrıca, split'ten farklı olarak dize sonundaki "\n" karakterini de göz ardı eder. Bir dosyanın içeriğini bütün olarak bir dizeye yüklediyseniz, onu satırlara bölmek için en uygunu splitlines kullanmaktır.

# Tuple

Stringler ile Tuple ortak özellikleri değiştirilemez (immutable) olmalarıdır. Elemanlarına yeni değerler atayamazsınız ve parantez ile gösterilir ()

In [87]:
a = (3.14, 2, 'Albert', 3+4j, (7.25,-23)) # Her virgülden sonrasını bir eleman olarak kabul eder. 
a[0]  # birinci eleman

3.14

In [88]:
a[-1][0] # Eleman kümesi içinde elaman oluşturmakta mümkündür. Bu örnekte gördüğümüz gibi son elemanın birinci elemanı

7.25

In [89]:
a[-1][1] # Son elemanın ikinci elemanı

-23

Çokuzlar atamada kullanılırsa, her eleman birebir eşleştirilerek atama yapılır.

In [90]:
(x,y,z)=[(5,-2.5), ("asd", 21), "Hello"]

In [91]:
z[0] 

'H'

In [92]:
y[0]

'asd'

In [93]:
y[0][2] # b ("asd") nin ikinci karakterini çektik.

'd'

In [94]:
a[0]+y[1] # a'nın ilk elemanı (3.14) ile y'nin ikinci elemanını (21) topladık

24.14

# Listeler

 Listeler köşeli parantez ile gösterilir []

In [95]:
liste = [5, "merhaba", (2,3,1), [-1,0,2]]
liste[0] = 4.1  # birinci elemanı değiştir
liste[-1][0] = 12 # son elemanın birinci elemanını değiştir
liste

[4.1, 'merhaba', (2, 3, 1), [12, 0, 2]]

In [96]:
liste[1]="merhabaDegisti" # İkinci elemana başka bir dize atadık.
liste

[4.1, 'merhabaDegisti', (2, 3, 1), [12, 0, 2]]

In [97]:
del liste[0] # İlk elemanı "5" sildik.
liste

['merhabaDegisti', (2, 3, 1), [12, 0, 2]]

Ancak listenin bir elemanı değiştirilemez tipteyse o elemanın alt elemanlarını silemez veya onlara atama yapamazsınız.

In [98]:
liste[0][2]="H"

TypeError: 'str' object does not support item assignment

In [99]:
isimlerListesi = ["ahmet", "mehmet", "veli", "ali"]  # Listelerde birleştirme yapabiliriz.
soyIsimlerListesi=["yılmaz", "deniz", "saygılı", "karaman"]
adSoyadListesi = isimlerListesi+soyIsimlerListesi
adSoyadListesi

['ahmet', 'mehmet', 'veli', 'ali', 'yılmaz', 'deniz', 'saygılı', 'karaman']

# Dictionary

Sözlüklerde her veri için bir anahtar kelime vardır. Bu veriyi çağırmak için anahtar kelime kullanılır. Süslü parantez ile gösterilir.


In [100]:
d={"selamVer": "Merhaba", "yas":27}

In [101]:
d["selamVer"]

'Merhaba'

In [102]:
d["yas"]

27

Sözlüklerde referans olarak sadece değiştirilemez tipler kullanılabilir (sayılar, dizeler, çokuzlar). Listeler kullanılamaz.

In [103]:
d[ [1,2,3] ] = 5

TypeError: unhashable type: 'list'

Sözlük metodları
------------------
Bir sözlük nesnesine ait metodları `dir` fonksiyonuyla görebiliriz. Başında ve sonunda altçizgi olan metodlar nesne içi kullanım içindir

In [104]:
dir(dict)

['__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__ior__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__ror__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'clear',
 'copy',
 'fromkeys',
 'get',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'update',
 'values']

### Sözlükteki bütün maddeleri silmek clear 

`Clear` metodu ile bir sözlükteki bütün anahtar-değer çiftleri silinebilir.

In [105]:
D = {'bir': 1, 'iki': 2.0, 'pi': 3.14159}
D.clear()
D

{}

Sözlüğün kopyasını çıkarmak
---------
Bazen bir sözlükteki orijinal bilgileri değiştirmeden, bir kopya üzerinde işlemler yapmanız gerekebilir. Bir sözlüğün kopyasını `copy` metoduyla elde edebilirsiniz.

In [106]:
D = {'bir': 1, 'iki': 2.0, 'pi': 3.14159}
D2 = D.copy()  # Bu işlemden sonra D2 de yapacağımız değişiklikler D'yi etkilemez.
D2

{'bir': 1, 'iki': 2.0, 'pi': 3.14159}

In [107]:
D2["bir"] = "uno"
print("D =",D)
print("D2 =",D2)

D = {'bir': 1, 'iki': 2.0, 'pi': 3.14159}
D2 = {'bir': 'uno', 'iki': 2.0, 'pi': 3.14159}


İki sözlüğü birleştirmek
-------------

Sözlüklerin `update` metodu ile varolan bir sözlüğe yeni bir sözlükten anahtar-değer çiftleri ekleyebilirsiniz. Aynı anahtara sahip iki değer varsa, eskisi yenisiyle değiştirilir. Bu metod nesne içi (in-place) değişiklik yapar.

In [108]:
telefonlar = {"Ali": 3762, "Zeynep": 8249, "Mehmet": 2652}
telf_2 = {"Fatma": 9212, "Deniz": 1129, "Mehmet": 1212}
telefonlar.update(telf_2)
telefonlar

{'Ali': 3762, 'Zeynep': 8249, 'Mehmet': 1212, 'Fatma': 9212, 'Deniz': 1129}

Elemanlara erişmek
-------------
Bir `D` sözlüğünde saklanan bir değere erişmenin en basit yolu `D[anahtar]` komutunu kullanmak. Fakat `anahtar` referansı sözlükte tanımlı değilse bir hata mesajı çıkar, bu da programınızın akışınızı bozabilir.

In [109]:
telefonlar = {"Ali": 3762, "Zeynep": 8249, "Mehmet": 2652}
telefonlar["Ali"]  # Sözlükde "Ali" ye karşılan gelen "3762" değerini çıktı olarak verdi

3762

In [110]:
telefonlar["Kaan"] # Sözlükte "Kaan" şeklinde bir anahtar kelime olmadığı için hata mesajı verdi.

KeyError: 'Kaan'

Verilen anahtarın sözlükte bulunmaması durumunda varsayılan bir değer döndürülmesini istiyorsanız, `get` metodunu kullanabilirsiniz. `D.get(k, d)` ifadesi, `k` anahtarı sözlükte varsa `D[k]` değerini, yoksa varsayılan `d` değerini geri verir.

In [111]:
for isim in ["Ali","Kaan"]:
    print("{} telefonu {}".format(isim, telefonlar.get(isim, "mevcut değil")))

Ali telefonu 3762
Kaan telefonu mevcut değil


Sözlüğe yeni anahtarlar eklemek
------------------------

Bir `D` sözlüğüne yeni bir anahtar-değer çiftini dinamik olarak kolayca ekleyebiliriz.

In [112]:
D = {}
D["abc"] = 3
D

{'abc': 3}

Bunun başka bir yolu da setdefault metodunu kullanmaktır. Bu metod, get gibi çalışır, mevcut olmayan bir anahtar verildiğinde bir varsayılan değer döndürür. Ama get'ten farklı olarak, verilen yeni anahtarı sözlüğe ekler.

In [113]:
print(D.get("def",42))
print(D) # sözlük eski halinde kalır.

42
{'abc': 3}


In [114]:
print(D.setdefault("def",42))
print(D) # sözlüğe "def" anahtarı eklendi.

42
{'abc': 3, 'def': 42}


### Dictionary: keys ve values

In [115]:
yeniSozluk= {"key1":100, "key2":[10,20,30,40,4.5,"yılmaz"],"key3":{"key4":4}} #Karmaşık ve iç içe bir sözlük oluşturduk

In [116]:
yeniSozluk.keys() # keys komutu ile oluşturulan sözlüğün anahtar kelimelerini görebiliriz.

dict_keys(['key1', 'key2', 'key3'])

In [117]:
yeniSozluk.values() # Values komutu sayesinde anahtar kelimelere ait değerleri (value) görebiliriz

dict_values([100, [10, 20, 30, 40, 4.5, 'yılmaz'], {'key4': 4}])

Bu sayede karmaşık sözlüklerde kullanmamız gereken anahtar kelimeleri ve sahip oldukları değerleri daha kolay görebiliriz.

In [118]:
yeniSozluk["key2"][-1]

'yılmaz'

In [119]:
yeniSozluk["key3"]["key4"]

4

## Metodlar

### Append

In [120]:
benimListem = [10,20,30,40,50]

In [121]:
benimListem

[10, 20, 30, 40, 50]

In [122]:
benimListem.append(100) # Append komutu ile listenin sonuna yeni eleman eklenebilir.
benimListem

[10, 20, 30, 40, 50, 100]

### Remove

In [123]:
benimListem.remove(50) # Verilen bir değere sahip olan ilk elemanı bulur ve onu listeden siler. 
                       # Değer listede yoksa ValueError hatası verir.
benimListem

[10, 20, 30, 40, 100]

### Pop

In [124]:
benimListem.pop() # Belli bir konumdaki elemanın değerini verir ve listeden siler. Parametre almazsa son elemanı siler.

100

In [125]:
benimListem

[10, 20, 30, 40]

### Count

In [126]:
benimListem.count(30) # Sorgulanan değerin listede kaç adet olduğunu gösterir.

1

In [127]:
benimListem.append(30)

In [128]:
benimListem

[10, 20, 30, 40, 30]

In [129]:
benimListem.count(30) # Az önceki sorguda bir değer bulurken bir tane daha 30 ekledik ve değerimiz ikiye çıktı

2

### Reverse

Liste elemanlarının sırasını kendi içinde (in-place) ters çevirir. DİKKAT: Yeni bir liste döndürmez, mevcut listeyi değiştirir. Eğer orijinal listeyi bozmak istemiyorsanız bir kopya çıkarmalısınız.

In [130]:
benimListem.reverse() # Listeyi tersine çevirir.
benimListem

[30, 40, 30, 20, 10]

### İnsert

In [131]:
benimListem.insert(1,10) # Listede verilen konuma bir eleman sokar, sonraki elemanları kaydırır.
benimListem

[30, 10, 40, 30, 20, 10]

### Copy

In [132]:
kopyaListe = benimListem.copy
kopyaListe()

[30, 10, 40, 30, 20, 10]

### Extend

In [133]:
yeniListe=[1, "abc", 3] 
benimListem.extend(yeniListe)  # Listenin sonuna bir liste ekleyerek genişletir.Bu işlem append komutundan farklıdır.
benimListem # Aynı işlemi append ile yaparsak L2 listesi kendi başına L1'in son elemanı olur.

[30, 10, 40, 30, 20, 10, 1, 'abc', 3]

In [134]:
benimListem=[30, 10, 40, 30, 20, 10]
liste2=[1,2,3]
benimListem.append(liste2)
benimListem

[30, 10, 40, 30, 20, 10, [1, 2, 3]]

### İndex

In [135]:
benimListem.index(40) #Belli bir değerin ilk olarak hangi konumda olduğunu verir. Değer bulunamazsa ValueError hatası çıkarır.

2

### sort

Bir dizi nesnesini alır ve nesne elemanlarının düzgün sıralar.

In [136]:
benimListem=[30, 10, 40, 30, 20, 10]
benimListem.sort()  #Sıralamadan sonra orijinal liste değişir.
benimListem

[10, 10, 20, 30, 30, 40]

sort metodu orijinal listede değişiklik yapar. Bu tür nesne içi (in-place) değişiklikler performansı biraz artırsa da, bazen orijinal listeyi bozmamak isteyebiliriz. O zaman iki seçeneğimiz var: Orijinal listenin bir kopyasını çıkarıp onu sıralarız:

In [137]:
benimListem=[30, 10, 40, 30, 20, 10]
kopyaListe=benimListem.copy()
kopyaListe.sort()
print("Asıl liste", benimListem)
print("Sıralı Liste",kopyaListe)

Asıl liste [30, 10, 40, 30, 20, 10]
Sıralı Liste [10, 10, 20, 30, 30, 40]


Veya orjinal listeyi bozmadan sıralamak için sorted kullanılır

 ### sorted
 
 Bir dizi nesnesini alır ve nesne elemanlarının düzgün sıralanmış olduğu bir liste döndürür.

In [138]:
benimListem=[30, 10, 40, 30, 20, 10]
benimListem.sort()  # Listeyi kendi içinde (in-place) sıralar. Sıralamadan sonra orijinal liste değişmez
benimListem

[10, 10, 20, 30, 30, 40]

In [139]:
sorted("merhaba")

['a', 'a', 'b', 'e', 'h', 'm', 'r']

In [140]:
sorted(("Ziya","Meral","Kaan","Fındık"))

['Fındık', 'Kaan', 'Meral', 'Ziya']

key parametresine verilen bir fonksiyon sıralamadan önce her elemana uygulanır; sonuç sıralaması bu fonksiyona göre belirlenir.

In [141]:
sorted([1, 3, -2, 4, -5, 6, 4], key=abs)    # mutlak değere göre sıralar

[1, -2, 3, 4, 4, -5, 6]

In [142]:
sorted( [(1,2), (0,2), (3,4), (2,-1)] , key=sum)  # çokuz elemanların toplamına göre sıralar

[(2, -1), (0, 2), (1, 2), (3, 4)]

### reverse

Sıralamayı tersine çevirir

In [143]:
benimListem.sort(reverse=True)  # sırayı ters çevirmek için reverse=True parametresi verilebilir.
benimListem

[40, 30, 30, 20, 10, 10]

### Türkçe Sıralama

Python 3 Unicode kullandığı için İngilizce dışı harfleri kullanırken zorluk çıkarmaz. Ama alfabetik sıralama yapmaya çalıştığımızda doğru cevabı alamayabiliriz.

In [144]:
sorted(("Ahmet","Şebnem","Mehmet","Ziya","İsmail","Ümit"))

['Ahmet', 'Mehmet', 'Ziya', 'Ümit', 'İsmail', 'Şebnem']

Türkçe alfabeye göre sıralama yapmak için yerellik ayarı yapmalıyız. Bunun için locale modülünü yüklemeli, ardından da sorted fonksiyonuna key parametresi olarak locale.strxfrm fonksiyonunu vermeliyiz.

In [145]:
import locale
locale.setlocale(locale.LC_ALL, ("tr", 'UTF-8'))
sorted(("Ahmet","Şebnem","Mehmet","Ziya","İsmail","Ümit"), key=locale.strxfrm)

['Ahmet', 'İsmail', 'Mehmet', 'Şebnem', 'Ümit', 'Ziya']

### range

In [146]:
benimListem

[40, 30, 30, 20, 10, 10]

In [147]:
range(10) 

range(0, 10)

In [148]:
list(range(10)) # 0 dan başlayarak 10'a kadar (10 dahil değil) bir nesne oluşturur. 

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

sort metodu orijinal listede değişiklik yapar. Bu tür nesne içi (in-place) değişiklikler performansı biraz artırsa da, bazen orijinal listeyi bozmamak isteyebiliriz. O zaman iki seçeneğimiz var: Orijinal listenin bir kopyasını çıkarıp onu sıralarız:

In [149]:
benimListem=[30, 10, 40, 30, 20, 10]
kopyaListe=benimListem.copy()
kopyaListe.sort()
print("Asıl liste", benimListem)
print("Sıralı Liste",kopyaListe)

Asıl liste [30, 10, 40, 30, 20, 10]
Sıralı Liste [10, 10, 20, 30, 30, 40]


In [150]:
list(range(2,10)) # Eğer başlangıç değeri belirtiilirse oradan başlayarak bitiş değerine kadar nesne oluşturur.

[2, 3, 4, 5, 6, 7, 8, 9]

In [151]:
list(range(2,10,3)) # En sona yazılan değer "3" 

[2, 5, 8]

range işlemi bir liste değil, dizinin başlangıç, bitiş ve adım bilgilerini barındıran bir nesne geri verir sadece. Bu sayede, range(1000000) gibi çok uzun görünen bir dizi bile bellekte ancak range(10) kadar yer kaplar. Elemanlar ve alt aralıklar gerektiğinde aritmetik olarak hesaplanır.

### len

Kapsayıcı bir nesnenin kaç nesne barındırdığını verir. Sıralı olmayan tiplerle de kullanılabilir.

In [152]:
len([1,2,3]) # Liste içerisindeki nesne sayısını verir.

3

In [153]:
len("Merhaba") # Dize içerisindeki nesnenin kaç nesnesi olduğunu verir.

7

In [154]:
len({"a":45,"b":4.5}) # Sözlük içerisindeki anahtar kelime-değer sayısını verir.

2

In [155]:
len([]) # Boş bir listenin uzunluğunu sıfır olarak gösterir.

0

### list

`list`fonksiyonu herhangi bir sıralı nesneyi bir liste nesnesine çevirmek için kullanılır.

In [156]:
s = "Merhaba"
list(s) #Dizeyi listeye çevirir

['M', 'e', 'r', 'h', 'a', 'b', 'a']

In [157]:
t=(1,2,3,4)
list(t) # Çokuzu listeye çevirir.

[1, 2, 3, 4]

`tuple`fonksiyonu ise bir sıralı diziyi çokuza çevirir.

In [158]:
s="merhaba"
tuple(s)

('m', 'e', 'r', 'h', 'a', 'b', 'a')

In [159]:
L = [1,2,3,4]
tuple(L)

(1, 2, 3, 4)

### any

Bir liste içide en az bir eleman mantıksal doğruya denkse `True`verir, yoksa `False`verir. Python'da sıfır, boş liste `[]`, boş dize `""` gibi yapıların da `False`sayıldığını; sıfır olmayan sayıların ve boş olmayan nesnelerin de `True`sayıldığını hatırlayın.

In [160]:
any([0,'',True]) # Burada sondaki "True" manktıksal doğruya denk olduğu için True ifadesini döndürdü

True

In [161]:
any([0,'',False]) # Elemanların hiçbiri mantıksal doğruya denk olmadığı için False ifadesini döndürdü.

False

In [162]:
any([0,"dolu",False])  # İkinci eleman "dolu" ifadesi mantıksal doğru olduğu için True ifadesini döndürdü.

True

### enumerate

Bir sıralı nesnenin elemanlarına sıra numarası atamak için kullanılır. Özellikle döngülerde, bir nesne üzerinden iterasyon yaparken yararlı olur.

In [163]:
list(enumerate("abc def"))

[(0, 'a'), (1, 'b'), (2, 'c'), (3, ' '), (4, 'd'), (5, 'e'), (6, 'f')]

In [164]:
index = 0
for numara in list(range(5,15)):
    print(f"güncel numara: {numara} güncel index: {index}")
    index = index + 1

güncel numara: 5 güncel index: 0
güncel numara: 6 güncel index: 1
güncel numara: 7 güncel index: 2
güncel numara: 8 güncel index: 3
güncel numara: 9 güncel index: 4
güncel numara: 10 güncel index: 5
güncel numara: 11 güncel index: 6
güncel numara: 12 güncel index: 7
güncel numara: 13 güncel index: 8
güncel numara: 14 güncel index: 9


In [165]:
for eleman in enumerate(list(range(5,15))):
    print(eleman)

(0, 5)
(1, 6)
(2, 7)
(3, 8)
(4, 9)
(5, 10)
(6, 11)
(7, 12)
(8, 13)
(9, 14)


In [166]:
for (index,numara) in enumerate(list(range(5,15))):
    print(numara)

5
6
7
8
9
10
11
12
13
14


### max
`max` fonksiyonu bir sıralı nesne veya bir dizi parametre alır, aralarından en büyük değerli olanını geri verir.

In [167]:
max(4,2,8,3,1,7)

8

In [168]:
L = [4,2,8,3,1,7]
max(L)

8

Dizelerden oluşan bir sıralı nesne verildiğinde max alfabetik sırada en ileride olanını döndürür.

In [169]:
max(["dfg","zxy","abc"])

'zxy'

In [170]:
max("merhaba")

'r'

key parametresi ile her elemana önceden uygulanacak bir fonksiyon belirleyebilirsiniz ve maksimum bu fonksiyonun sonucuna göre tespit edilir. Söz gelişi, mutlak değer olarak en büyük elemanı bulmak için key=abs verebilirsiniz.

In [171]:
L = [-1,-3, 4, -5, 2] # abs = mutlak değer
max(L, key=abs)  # Burada her her elemanın mutlak değerini alıp olan göre maksimum değeri bulmasını istedik.

-5

### min
min fonksiyonu verilen bir dizinin içinde, veya parametreler içinde en küçük olanını döndürür. Kullanımı max ile aynıdır.

In [172]:
L = [-1,-3, 4, -5, 2]
min(L, key=abs)

-1

In [173]:
min("merhaba")

'a'

### sum

Bir sıralı nesnenin elemanlarının toplamını verir. Elemanlar sayısal değere sahip olmalıdır.

In [174]:
sum([1,2,3,4])

10

In [175]:
sum(range(1,1000)) # 1'den başlayarak 999' a kadar olan bütüm sayıların toplamını verdi.

499500

### zip

Aynı uzunlukta iki veya daha çok listenin aynı konumdaki elemanlarını sırayla alır, bu elemanlardan çokuzlar oluşturur ve bu çokuzların listesini verir.

In [176]:
L1 = [4, 5, 9]
L2 = ['a', 'b', 'c']
list(zip(L1, L2))

[(4, 'a'), (5, 'b'), (9, 'c')]

`zip` ile farklı verileri birleştirerek bir for döngüsünde işlemek mümkün olur. Tıpkı `enumerate` gibi zip de bizi elemanları bir sayaç değişkeniyle takip etme mecburiyetinden kurtarır. 

In [177]:
yaşlar = [19, 25, 32, 27]
boylar = [170, 180, 175, 169]
ağırlık = [75, 78, 81, 71]
for y, b, a in zip(yaşlar, boylar, ağırlık):
    print("Yaş:", y,"Boy:", b, "Ağırlık:", a)

Yaş: 19 Boy: 170 Ağırlık: 75
Yaş: 25 Boy: 180 Ağırlık: 78
Yaş: 32 Boy: 175 Ağırlık: 81
Yaş: 27 Boy: 169 Ağırlık: 71


### Hataya dikkat: Nesne içi değişiklik yapan metodlar
Yukarıda gördüğümüz sort ve reverse gibi metodlar, bağlı oldukları listenin verisini doğrudan doğruya değiştirirler. Buna nesne içi (in place) değiştirme denir. Geriye döndürdükleri bir değer yoktur.

Yukarıda gördüğümüz sort ve reverse gibi metodlar, bağlı oldukları listenin verisini doğrudan doğruya değiştirirler. Buna nesne içi (in place) değiştirme denir. Geriye döndürdükleri bir değer yoktur.

In [178]:
benimListem=[30, 10, 40, 30, 20, 10]
print(benimListem.sort())

None


Yeni başlayan programcıların yaptıkları yaygın bir hata benimListem.sort() çağrısını bir atamanın sağ tarafında kullanmaktır.

In [179]:
L2=benimListem.sort()

Programcı burada L2'nin L'nin sıralanmış hali olduğunu ummaktadır (aslında L2 = sorted(L) yazması gerekirdi). Python bu atamayı sessizce yapar, ama aslında L2'yi bir listeye değil None değerine atamıştır.

In [180]:
L2 is None

True

Programın içinde daha sonra L2'yi kullanmak isteyen programcı şaşırtıcı hata mesajlarıyla karşılaşabilir.

In [181]:
L2[0]

TypeError: 'NoneType' object is not subscriptable

## Format = Biçimlendirme


Birçok durumda bir değişkende barındırılan bir değeri uygun bir biçimde düzenleyerek ekrana basmamız gerekir. Bu düzenlemeler ve biçimlendirmeler örneğin şöyle olabilirler:
* Bir kalıp dize içindeki boşlukları doldurmak, `İsim:...., Yaş:....` gibi.
* Sayıları ondalıktan sonra sabit sayıda basamakla vermek: `9.340, -18.731` gibi.
* Sayıları sıfırlarla doldurmak: `099.45, 001.12` gibi.
* Tablo halinde basılan listeleri hizalı olarak, belli boşluklar ayırarak basmak.
* Tablo içeriklerini sağa veya sola yaslamak, ortalamak.

Bu tür düzenlemeler için Python'da kapsamlı bir mini-dil mevcut. Bu tür düzenlemeler, `format()` metoduyla beraber şu şekilde kullanılır:

    "<kalıp dizesi>".format(<değerler>)

Dize biçimlendirmenin bütün ayrıntıları için [Python belgelerine](https://docs.python.org/3/library/string.html#formatstrings) bakabilirsiniz.
    
Konumla yerleştirme ve isimle yerleştirme
------------------------
En basit kullanımda, `format()`'a verilen değerler, kalıp dizesinin içinde küme parantezleri (`{}`) ile bırakılan boşluklara sırayla yerleştirilir.

In [182]:
"İsim: {}, Telefon: {}, Boy: {} m".format("Kaan", "5555", 1.80)

'İsim: Kaan, Telefon: 5555, Boy: 1.8 m'

Yukarıdaki örnekte birinci boşluğa birinci değer (`"Kaan"`), ikinci boşluğa ikinci değer (`"5555"`), üçüncü boşluğa da üçüncü değer (1.80) yerleştirildi. Bu sırayı değiştirmek için, küme parantezlerinin içine arzu ettiğimiz konum numaralarını koyabiliriz.

In [183]:
"İsim: {0}, Boy: {2} m, Telefon: {1}".format("Kaan", "5555", 1.80)

'İsim: Kaan, Boy: 1.8 m, Telefon: 5555'

Başka bir yol da, yer tutuculara isim vermek ve `format()`'a bu isimlerle parametreler vermektir.

In [184]:
"Alıcı Adı: {isim}, Yaşı: {yaş}, Şehir: {şehir}".format(yaş=32, isim="Fatma", şehir="Tekirdağ")

'Alıcı Adı: Fatma, Yaşı: 32, Şehir: Tekirdağ'

Örneklerden gördüğümüz gibi, `format()` fonksiyonu kalıp dizesindeki boşluklarla değerleri eşleştirirken, fonksiyonlardaki [parametre eşleştirme](http://www.veridefteri.com/2018/01/10/python-programlamaya-giris-fonksiyon-parametreleri/) kurallarını kullanıyor. İlk gördüğümüz örnek _konumla eşleştirmeye_ denk iken, son örneğimiz _isimle eşleştirme_ yapıyor.

Bu sebepten, fonksiyon çağrılarındaki _parametre çözme_ kuralları da aynen uygulanabilir. Mesela:

In [185]:
p = ("Kaan","5555",1.80)
"İsim: {}, Telefon: {}, Boy: {} m".format(*p)

'İsim: Kaan, Telefon: 5555, Boy: 1.8 m'

`format()`'a verilen parametre bir liste veya sözlük ise, kalıp dizesinde indeksleme uygulanabilir.

In [186]:
listem = [3,6,7,9]
"İlk eleman: {0[0]}, son eleman: {0[3]}".format(listem)

'İlk eleman: 3, son eleman: 9'

In [187]:
kayıt = {"isim":"Kaan", "boy":1.80, "telefon":"5555"}
"İsim: {D[isim]}, Tel: {D[telefon]}, Boy: {D[boy]}".format(D=kayıt)

'İsim: Kaan, Tel: 5555, Boy: 1.8'

Sayı biçimlendirme
------------------
Yer tutucu içine bir sayı koyacağımız zaman, tamsayılarda `{:d}`, ondalıklı sayılarda ise `{:f}` kodlarını kullanırız.

In [188]:
"Yaş: {:d}, Boy: {:f}".format(42, 1.76) 

'Yaş: 42, Boy: 1.760000'

Belli miktarda, söz gelişi 5 karakterlik yer ayırmak için `{:5d}` yazımını kullanabiliriz. Ondalıklı sayılar için `{:5.2f}` yazımı, noktadan sonra 2 basamak olmak üzere en az 5 karakterlik yer ayrılmasını sağlar (ondalık noktası dahil). Türkçeye uygun olarak virgüllerle ayırma yöntemi için bir sonraki bölüme bakın.

In [189]:
"Yaş: {:5d}, Boy: {:5.3f}".format(42, 1.76) 

'Yaş:    42, Boy: 1.760'

Ayırdığımız alanda, sayının değerini değiştirmeden boşlukların sıfırlarla doldurulmasını istiyorsak `{:05d}` veya `{05.3f}` gibi yazımlar kullanırız.

In [190]:
"Yaş: {:05d}, Boy: {:06.3f}".format(42, 1.76) 

'Yaş: 00042, Boy: 01.760'

Negatif sayıların başına her zaman eksi gelir; pozitif sayıların da her zaman artı ile başlaması için `{:+d}` yazımını kullanırız.

In [191]:
"{:+d}, {:+d}".format(42, -45)

'+42, -45'

Bir sayıyı ikili, sekizli, onaltılı tabanda göstermek için sırasıyla `b`, `o`, `x` kullanırız.

In [192]:
print("Onluk tabanda {:d}\nİkili tabanda {:b}\nSekizli tabanda {:o}\nOnaltılı tabanda {:x}".format(42,42,42,42))

Onluk tabanda 42
İkili tabanda 101010
Sekizli tabanda 52
Onaltılı tabanda 2a


Ondalıklı sayıları göstermek için şu kodlar kullanılabilir:
* `{:f}` veya `{:F}` (fixed): Ondalıktan sonra sabit sayıda basamak (varsayılan 6).
* `{:e}` veya `{:E}` (exponential): Sayıyı bilimsel notasyonda gösterir; üsteli `"e"` veya `"E"` harfinden sonra koyar.
* `{:g}` veya `{:G}` (genel): Yerine göre `f` veya `e`. Belirli bir basamak sayısını (varsayılan 6) verecek şekilde yuvarlar; sayının büyüklüğüne göre `f` veya `e` biçimine getirir.
* `{:%}`: Sayıyı 100 ile çarpar ve sonuna yüzde işareti koyar.

In [193]:
x = 12345.6789
print("Sabit biçim: {:f}, {:F}".format(x,x))
print("Üstel biçim: {:e}, {:E}".format(x,x))
print("Genel biçim: {:g}, {:G}".format(x,x))

Sabit biçim: 12345.678900, 12345.678900
Üstel biçim: 1.234568e+04, 1.234568E+04
Genel biçim: 12345.7, 12345.7


In [194]:
x = 12345.6789
print("Sabit biçim: {:4.2f}, {:4.4F}".format(x,x))
print("Üstel biçim: {:1.3e}, {:1.4E}".format(x,x))
print("Genel biçim: {:.10g}, {:.2G}".format(x,x))

Sabit biçim: 12345.68, 12345.6789
Üstel biçim: 1.235e+04, 1.2346E+04
Genel biçim: 12345.6789, 1.2E+04


Yüzdeleri gösterirken `%` kodunu kullanabiliriz.

In [195]:
geçen = 57
başvuran = 245
"Başarı oranı {:.2%}".format(geçen/başvuran)

'Başarı oranı 23.27%'

Sayıları Türk standartlarında biçimlendirmek
-----------------
Yukarıdaki örneklerde, Amerikan standardına uygun olarak, sayı grupları virgülle ayrıldı ve ondalık için nokta kullanıldı. Oysa Türkiye'deki sayı yazma standartlarında sayı grupları noktayla ayrılır, ondalık için ise virgül kullanılır. Türk standartlarına uygun biçimlendirme yapmak için biraz dolaylı bir yoldan gidip, yerelleştirme işlemleri için kullanılan `locale` kütüphanesini kullanmamız gerekir.

In [196]:
import locale
loc = locale.getlocale()
locale.setlocale(locale.LC_ALL,"tr_TR.UTF-8")

'tr_TR.UTF-8'

Yerelleştirme ayarlarını yapsak da, bu ayarlar `str.format` metodunu etkilemeyecektir. Bunun yerine, `locale.format` fonksiyonunu kullanmalıyız. Bu fonksiyonun dize biçimleme sintaksı [eski usül](https://docs.python.org/2/library/stdtypes.html#string-formatting) olarak bilinir. Basit bir örnek olarak:

In [197]:
locale.format("%f", 1234567.89, grouping=True)

C:\Users\Administrator\AppData\Local\Temp\ipykernel_11780\742679681.py:1: DeprecationWarning: This method will be removed in a future version of Python. Use 'locale.format_string()' instead.
  locale.format("%f", 1234567.89, grouping=True)


'1.234.567,890000'

Şimdi de virgülden sonra 3 hane olmak üzere toplam 13 karakterlik yer ayırarak biçimlendirelim.

In [198]:
locale.format("%013.3f", 1234567.89, grouping=True)

C:\Users\Administrator\AppData\Local\Temp\ipykernel_11780\3145349833.py:1: DeprecationWarning: This method will be removed in a future version of Python. Use 'locale.format_string()' instead.
  locale.format("%013.3f", 1234567.89, grouping=True)


'001.234.567,890'

Nesne içi değişkenleri kullanmak
---------
Nesnelerin iç değişkenleri ("attributes") kalıp dizede kullanılabilir. Örneğin, karmaşık sayıların gerçek kısımları `real`, sanal kısımları `imag` isimli iç değişkenlerde saklanır.

In [199]:
z = 3+2j+5
z.real, z.imag # real sayı 3+5 = 8 dir. Sanal kısım ise 2j den gelir.

(8.0, 2.0)

Dize biçimlendirmede bu iç değişkenlerin isimlerini doğrudan kullanabiliriz.

In [200]:
"Gerçek kısım {0.real}, sanal kısım {0.imag}".format(3+2j+5)

'Gerçek kısım 8.0, sanal kısım 2.0'

## Küme Metodları



Bir küme, listelerden farklı olarak, belli bir sırası bulunmayan bir veri topluluğudur; yani elemanlarına indeksleme ile erişilemez. Ayrıca, matematiksel kümeler gibi, aynı elemandan birden fazla barındıramaz. Python'da bir küme yaratmanın birinci yolu, nesneleri küme parantezinde toplayarak vermektir:

In [201]:
s = {(1,2,3),"abc",123,-1,"abc",(1,2,3)}
s

{(1, 2, 3), -1, 123, 'abc'}

İkinci yolu ise, `set()` fonksiyonuna bir sıralı nesne vermektir:

In [202]:
set([1,3,1,-2,"abc", "abc"])

{-2, 1, 3, 'abc'}

Bir dize de karakterlerden oluşan bir sıralı nesnedir, bu yüzden `set()` fonksiyonu ile dizenin karakterlerinin kümesini elde ederiz.

In [203]:
set("ananas")

{'a', 'n', 's'}

Örneklerden de gördüğünüz gibi kümelerde, matematiksel tanıma uygun olarak, bir eleman sadece bir kere mevcut olabilir. Orijinal sıralı nesnede mükerrer eleman varsa alınmaz.

Python'da bir liste, bir kümenin elemanı olamaz. Denemeye kalkarsanız hata mesajı alırsınız.

In [204]:
{1,2, [1,2,3]}

TypeError: unhashable type: 'list'

Bir kümedeki elemanların _hashable_ olabilmesi, yani bir hash fonksiyonuna verilebilmesi gerekir. Bu da genellikle değiştirilemez (immutable) veri tipleri olmalarını gerektirir. Sayılar, çokuzlar, dizeler küme elemanı olmaya uygundurlar. Buna karşılık listeler, sözlükler ve kümeler, küme elemanı olamazlar (evet, matematiksel tanımda kümeler başka kümelerin elemanı olabilir, ama Python'da teknik sebeplerle bu mümkün olmaz).

Boş küme için `set()` fonksiyonu tek başına kullanılmalıdır. Boş küme parantezleri `{}` kullanmak boş bir sözlük oluşturur, küme değil.

In [205]:
type(set()), type({})

(set, dict)

Bir kümenin kaç elemanlı olduğunu bulmak için `len()` fonksiyonu kullanılabilir.

In [206]:
s = {1,3,1,2,1,3,2}
len(s)

3

Bir elemanın kümenin içinde olup olmadığını sorgulamak için `e in s` kullanılabilir. Ters şart olarak da `e not in s` kullanılabilir.

In [207]:
1 in s, 4 in s, 4 not in s

(True, False, True)

Küme karşılaştırmaları
---
Küme nesneleri `==`, `<`, `>` gibi karşılaştırma işlemlerini yeniden tanımlayarak küme teorisine uyumlu hale getirirler. Söz gelişi `A <= B` şartı, A kümesi B kümesinin bir alt kümesi ise doğrudur.

In [208]:
Avrupa = {"Almanya","Fransa","İspanya","Portekiz","Rusya","Türkiye"}
Asya = {"Rusya","Türkiye","Çin","Hindistan"}
İberya = {"İspanya","Portekiz"}

Eşitlik (`==`) ve eşitdeğil (`!=`) işlemleri iki kümenin eşit olup olmadığını test eder.

In [209]:
Avrupa == Avrupa, Avrupa != Asya

(True, True)

Alt küme testi için `<=` işlemi, öz alt küme testi için `<` işlemi kullanılabilir.

In [210]:
İberya < Avrupa, Asya < Asya, Asya <= Asya

(True, False, True)

Kapsama işlemi için `>` ve `>=` işlemleri kullanılır.

In [211]:
Avrupa > İberya, Asya > Asya, Asya >= Asya

(True, False, True)

Bileşim, kesişim, fark, simetrik fark
----

İki kümenin bileşim kümesini `|` işlemi ile elde ederiz.

In [212]:
Avrupa | Asya

{'Almanya',
 'Fransa',
 'Hindistan',
 'Portekiz',
 'Rusya',
 'Türkiye',
 'Çin',
 'İspanya'}

Kümelerin kesişim kümesi ise `&` işlemi ile elde edilir.

In [213]:
Avrupa & Asya

{'Rusya', 'Türkiye'}

Fark işlemi `A-B`, `A` kümesinde olup `B` kümesinde olmayanları verir.

In [214]:
Avrupa - İberya

{'Almanya', 'Fransa', 'Rusya', 'Türkiye'}

Simetrik fark `A^B` işlemi, ya `A` kümesinde ya da `B` kümesinde olan, ama ikisinde birden olmayan elemanları verir. Bu işlem `A^B == (A|B) - (A&B)` eşitliğini sağlar.

In [215]:
Avrupa ^ Asya

{'Almanya', 'Fransa', 'Hindistan', 'Portekiz', 'Çin', 'İspanya'}

Donuk kümeler
--
Python, kümelere alternatif olarak bir de _donuk küme_ (frozen set) nesnesi sunar. İkisi arasındaki tek fark, donuk kümelerin bir kere tanımlandıktan sonra değiştirilemez (immutable) olmasıdır. Yani donuk kümeye eleman eklenip çıkarılamaz, kümede değişiklik yapılamaz. Değiştirilemezlik sayesinde donuk kümeler _hashable_ nesnelerdir. Böylece sözlüklerde anahtar olarak kullanılabilirler, veya küme elemanı olabilirler.

In [216]:
s = frozenset((1,2,"abc",frozenset("abc")))
s

frozenset({1, 2, 'abc', frozenset({'a', 'b', 'c'})})

Buraya kadar gösterdiğimiz bütün işlemler, karşılaştırmalar, bileşim, kesişim ve fark işlemleri donuk kümelerde de aynen kullanılabilir. Ancak buradan sonra anlatılacak işlemler kümelerde değişiklik yapmak için kullanıldığından, sadece normal (donuk olmayan) kümeler için geçerlidirler.

Kümelere eleman ekleme ve çıkarma
----
Bir kümeye tek tek eleman eklemek için `add()` metodunu kullanırız.

In [217]:
s = set()

s.add(1)
s.add("abc")
s.add(2)
s.add(1)

s

{1, 2, 'abc'}

Mevcut bir elemanı kümeden silmek için ise `remove()` metodu kullanılabilir.

In [218]:
s.remove(2)
s

{1, 'abc'}

Ancak, belirtilen eleman kümede yoksa, `remove()` çağrısı bir `KeyError` hatası verir.

In [219]:
s.remove(2)

KeyError: 2

Bu hata mesajını yakalayıp ona göre bir işlem yapmanız gereken durumlar olabilir. Eğer böyle bir niyetiniz yoksa `discard()` metodunu kullanabilirsiniz. Bu işlem de bir elemanı kümeden kaldırır, ama kümede o eleman mevcut değilse hata mesajı vermez.

In [220]:
s.discard(2)

Bir kümedeki bütün elemanları silmek için `clear()` metodunu kullanırız.

In [221]:
s.clear()
s

set()

Bir kümeden herhangi bir elemanı döndüren ve o elemanı kümeden çıkaran bir işlem için `pop()` metodu kullanılabilir. Kümede eleman kalmadıysa, `pop()` metodu `KeyError` hatası verir.

In [222]:
s = {1,"abc",(1,2,3)}
s.pop(), s.pop(), s.pop()

(1, 'abc', (1, 2, 3))

In [223]:
s.pop()

KeyError: 'pop from an empty set'

Küme güncelleme işlemleri
---
Sayılarda ve listelerde `+=`, `*=` gibi güncelleme işlemleri olduğunu görmüştük. Kümelerde de, yukarıda gördüğümüz bileşim, kesişim, fark, simetrik fark işlemlerini `|=`, `&=`, `-=`, `^=` yapılarıyla güncelleme işlemi haline getirmek mümkün.

Kesişim kümesi boş olmayan iki küme üretelim:

In [224]:
s1 = set(range(5))
s2 = set(range(2,7))
s1, s2

({0, 1, 2, 3, 4}, {2, 3, 4, 5, 6})

Şimdi `s1` kümesini `s1` ve `s2`'nin bileşim kümesi olacak şekilde güncelleyelim:

In [225]:
s1 |= s2  # s1 = s1 | s2 ile aynı
s1,s2

({0, 1, 2, 3, 4, 5, 6}, {2, 3, 4, 5, 6})

`s1`'i eski haline getirip, `s1` ile `s2`'nin kesişim kümeleri olacak şekilde güncelleyelim:

In [226]:
s1 = set(range(5))
s1 &= s2  # s1 = s1 & s2 ile aynı
s1

{2, 3, 4}

Aynısını fark ve simetrik fark işlemleriyle yapalım:

In [227]:
s1 = set(range(5))
s1 -= s2  # s1 = s1 - s2 ile aynı
s1

{0, 1}

In [228]:
s1 = set(range(5))
s1 ^= s2   # s1 = s1 ^ s2 ile aynı
s1

{0, 1, 5, 6}

## Random

In [229]:
from random import randint

In [230]:
randint(0,100)

45

In [231]:
randint(0,100)

29

In [232]:
yeniListe = list(range(0,10))
yeniListe

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [233]:
from random import shuffle
shuffle(yeniListe)
yeniListe

[6, 0, 1, 7, 3, 5, 8, 2, 4, 9]

Uygulamalar
--
<h3>Mükerrer (tekrarlanan) elemanları kaldırmak</h3>
Bir listeyi kümeye çevirmek mükerrer elemanların atılmasını sağlar. Tekil elemanların listesini bulmak için bu özelliği kullanabiliriz.

In [234]:
L = [1,2,3,4,1,2,3,4,1,3,2,1,4,1,2,1,3,1]
list(set(L))

[1, 2, 3, 4]

Elbette aynı yöntemi çokuzlar ve dizelerle de uygulayabiliriz.

In [235]:
dize = "korkma sönmez bu şafaklarda yüzen al sancak"
"".join(sorted(set(dize))) # Boş bir dizenin içine "dize" olarak tanımlanan değişkenin içerisindeki dizeyi aktardık.
                           # Bunu yaparken sorted ile küçükten büyüğe sıraladık ve 'set' komutunu kullandığmıız için 
                           # Aynı olan karakterleri sadece bir kere aldı

' abcdefklmnorsuyzöüş'

<h3>Elemanın varlığını test etmek</h3>
Bir kümede belli bir elemanın mevcudiyetini test etmek, listede test etmekten çok daha hızlıdır. Bunun sebebi, bir eleman aranırken listenin baştan sona taranmasıdır. Bu yüzden, liste uzadıkça listede bir elemanın varlığının tespiti için gereken zaman artar. Kümelerde ise hash fonksiyonu kullanarak bir tek işlemde test yapılabilir; işlem kümedeki eleman sayısına bağlı değildir.

Bu hız farkını görmek için 1-1.000.000 arasındaki sayılardan oluşan bir liste ve bir küme oluşturalım. Mükerrer eleman olmadığı için ikisi de aynı uzunlukta olacak.

In [236]:
L = list(range(1,1000001))
s = set(L)

In [237]:
from random import randint

1 ile 2.000.000 arasında rastgele bir sayı seçelim ve sayının listede ve kümede varlığını test edelim. Yüzde elli ihtimalle sayı mevcut olmayacak. `%%timeit` hücre sihri ile bu işlemi defalarca tekrarlayıp ortalama süresini bulalım.

In [238]:
%%timeit
n = randint(1,2000000)  # Liste içinde arama yapıyoruz.
n in L

5.51 ms ± 246 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [239]:
%%timeit
n = randint(1,2000000)  # Küme içinde arama yapıyoruz.
n in s

519 ns ± 5.61 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Küme içinde aramanın listede aramaya göre bin kat daha hızlı olduğunu görebiliyoruz. 

Şimdi aynı işlemi bir sözlük ile yapalım. Anahtar değerleri 1-1.000.000 arasında olan büyük bir sözlük yaratalım ve yine rastgele seçilmiş anahtar değerlerini sorgulayalım.

In [240]:
D = dict.fromkeys(L)

In [241]:
%%timeit
n = randint(1,2000000)
n in D

560 ns ± 11.4 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Sözlüklerde eleman tespiti kümelerdeki kadar hızlıdır, çünkü sözlüklerde de anahtarlar hash fonksiyonlarıyla aranır. Bütün sözlük elemanlarını taramak gerekmez. Yaptığımız zaman ölçümü de bunu gösteriyor.

# Boolean

Boolean iki farklı değer döndüren bir mantıksal veri tipidir. Ya 'true' ya da 'false' değerini döndürür. Yani atandığı değişken veya istenen işlemin sonucunda bu değerlerden birini verir.

In [242]:
10 > 5

True

In [243]:
10 < 5

False

In [244]:
listem = [5000,10000,3000,1000,2000,4000]
ortalama = sum(listem)/len(listem) # listemizin ortalamasını aldık sonuç = 4166.666
listem[3]>ortalama # listemizin üçünü elemanı (1000) ortalamadan büyük mü diye kontrol ettik.

False

In [245]:
kullaniciMaas = int(input("maaş bilgisini giriniz: "))
kullaniciMaas>ortalama

maaş bilgisini giriniz: 17000


True

# Mantıksal İşlemler

## and, or, not, ==, !=

`and`ve anlamnı taşır kendinden sonra gelen ifadenin de doğru olurse True döndürür.
`or`veya anlamı taşır ifadelerden herhangi birisi doğru olsa bile True döndürür.
`not`sorgulanan ifade yanlışsa True döndürür.
`==`sorgulanan değerler eşitse True döndürür.
`!=`sorgulanan ifade eşit değilse True döndürür

In [246]:
x=8
y=10

In [247]:
y==x

False

In [248]:
x!=y

True

In [249]:
x < y and x != y # Her iki şartta sağlandığı için True

True

In [250]:
x > y or x > 5 # şartlardan birisi sağlandığı için (x>5) True

True

In [251]:
1 > 3 and 3 > 2 # bir üçten büyük olmadığı için False

False

In [252]:
not 5 == 4 # Burada 5==4 sorgusunun False döndürmesi lazım fakat başına "not" koyarak sorguladığmız ifadenin yanlış olması 
           # durumunda True döndürmesini sağladık.

True

In [253]:
5 == 4 # Yukarıdaki ifaden farklı olarak burada "not" olmadığı için False döndürür.

False